merlin-framework/src/main/java/gov/nasa/jpl/aerie/merlin/framework/ReplayingTask.java

```diff
@@ -1,71 +1,227 @@
-      final var returnValue = this.task.get();
+    private int memoryIndex = 0;
+    private int ignoredIndex = 0;
 
-      // If we get here, the activity has completed normally.
-      return TaskStatus.completed(returnValue);
-    } catch (final Yield ignored) {
-      // If we get here, the activity has suspended.
-      return Objects.requireNonNull(handle.status, "Task status is null, but it should have been assigned a value on yield.");
+    public void rewind() {
+      this.memoryIndex = 0;
+      this.ignoredIndex = 0;
+    }
+
+    public <T> T remember(final Supplier<T> action) {
+      T value;
+      if (this.memoryIndex < this.memory.size()) {
+        // SAFETY: Tasks are deterministic, and last time we did this action, we cached a value of type `T`.
+        @SuppressWarnings("unchecked")
+        final var tmp = (T) this.memory.get(this.memoryIndex);
+        value = tmp;
+      } else {
+        value = action.get();
+        this.memory.add(value);
+      }
+
+      this.memoryIndex += 1;
+      return value;
+    }
+
+    public void forget(final Runnable action) {
+      if (this.ignoredIndex < this.ignoredInputs) {
+        // Do nothing
+      } else {
+        action.run();
+        this.ignoredInputs += 1;
+      }
+
+      this.ignoredIndex += 1;
     }
   }
 
-  private final class ReplayingTaskHandle implements TaskHandle {
-    public TaskStatus<Return> status = null;
+  /**
+   * A control-flow exception for quickly escaping from a task which cannot yet proceed any further.
+   *
+   * <p> This exception extends Error instead of RuntimeException to reduce the likelihood that
+   * it gets spuriously caught by an over-broad catch clause. </p>
+   */
+  private static final class Yield extends Error {
+    public static final Yield INSTANCE = new Yield();
+
+    public Yield() {
+      super(null, null, /* capture suppressed exceptions? */ false, /* capture stack trace? */ false);
+    }
+  }
+
+  private static final class TaskContext<Output> implements Context {
+    private final Function<Context, Output> task;
+    private final History history;
+
+    private final Scheduler scheduler;
+    private TaskStatus<Output> status = null;
+    private boolean isYielding = false;
+
+    public TaskContext(final Function<Context, Output> task, final Scheduler scheduler, final History history) {
+      this.task = Objects.requireNonNull(task);
+      this.scheduler = Objects.requireNonNull(scheduler);
+      this.history = Objects.requireNonNull(history);
+    }
+
+    public TaskContext(final Function<Context, Output> task, final Scheduler scheduler) {
+      this(task, scheduler, new History());
+    }
 
-    private Scheduler yield(final TaskStatus<Return> status) {
-      this.status = status;
-      throw Yield;
+    public static <Output>
+    TaskStatus<Output> replay(
+        final History history,
+        final Scheduler scheduler,
+        final Function<Context, Output> task
+    ) {
+      history.rewind();
```

Comment:
I don't think so. Since we create a new `TaskContext` every time we step the task, we're going to throw away the one that has `isYielding` stuck true anyway.